GitCommands/Git/Commands/GitCheckoutRevisionCmd.cs

```diff
@@ -0,0 +1,31 @@
+﻿using GitExtUtils;
+using GitUIPluginInterfaces;
+
+namespace GitCommands.Git.Commands
+{
+    public sealed class GitCheckoutRevisionCmd : GitCommand
+    {
+        public GitCheckoutRevisionCmd(
+            ObjectId revision,
+            LocalChangesAction localChanges = LocalChangesAction.DontChange)
+        {
+            Revision = revision;
+            LocalChanges = localChanges == LocalChangesAction.Stash ? LocalChangesAction.DontChange : localChanges;
+        }
+
+        public override bool AccessesRemote => false;
+        public override bool ChangesRepoState => true;
+        public LocalChangesAction LocalChanges { get; }
+        public ObjectId Revision { get; }
+
+        protected override ArgumentString BuildArguments()
+        {
+            return new GitArgumentBuilder("checkout")
+            {
+                { LocalChanges == LocalChangesAction.Merge, "--merge" },
+                { LocalChanges == LocalChangesAction.Reset, "--force" },
+                Revision.ToString().QuoteNE()
+            };
+        }
+    }
```

Comment:
Too much boilerplate code IMO. Do you foresee any use of the stored ctor arguments?
```suggestion
        public GitCheckoutRevisionCmd(
            ObjectId revision,
            LocalChangesAction localChanges = LocalChangesAction.DontChange)
            : base(accessesRemote: false, changesRepoState: true, BuildArguments())
        {
            return;

            ArgumentString BuildArguments()
            {
                return new GitArgumentBuilder("checkout")
                {
                    { localChanges == LocalChangesAction.Merge, "--merge" },
                    { localChanges == LocalChangesAction.Reset, "--force" },
                    revision.ToString().QuoteNE()
                };
            }
        }
    }
```
`BuildArguments()` can even be omitted and `new GitArgumentBuilder(...)` can be written in place.
I kept it because I expected that `LocalChanges = localChanges == LocalChangesAction.Stash ? LocalChangesAction.DontChange : localChanges;` had an effect. I do not see any.

Reply:
I'm not against your suggestion per se, but it requires changes to the base class and restructure of the other implementation. This is a bigger work that I currently have time for.

Reply:
There is no need for a class `GitCheckoutBranchCmd` anyway, a factory method would suffice:
```
        public static IGitCommand CreateGitCheckoutRevisionCmd(
            ObjectId revision,
            LocalChangesAction localChanges = LocalChangesAction.DontChange)
```

---

> This is a bigger work that I currently have time for.

I could prepare a commit.

Reply:
> There is no need for a class `GitCheckoutBranchCmd` anyway, a factory method would suffice:

I'm not against this either, but this should be done for all commands. We already have two different ways of "git command" implementation, and I'd like to avoid creating a third way.

> > This is a bigger work that I currently have time for.
> 
> I could prepare a commit.

Sure, if you have cycles. Thank you.

Reply:
I have simplified GitCommand in a first PR, rebased this one and gone a further step back to a common `GitCommandFactory.Checkout()`: see branch [mstv/convert_GitCommandHelpers_CheckoutCmd](https://github.com/mstv/gitextensions/tree/convert_GitCommandHelpers_CheckoutCmd)

Reply:
rebased and updated in my fork